virttest/utils_net.py

```diff
@@ -2379,6 +2381,55 @@ def cleanup(self):
         self.close_session()
 
 
+def ieee_eui_generator(base, mask, start=0, repeat=False):
+    """
+    IEEE extended unique identifier(EUI) generator.
+
+    :param base: The base identifier number.
+    :param mask: The mask to calculate identifiers.
+    :param start: The ordinal number of the first identifier.
+    :param repeat: Whether use repeated identifiers when exhausted.
+
+    :return generator: The target EUI generator.
+    """
+    offset = 0
+    while True:
+        out = base + ((start + offset) & mask)
+        yield out
+        offset += 1
+        if offset > mask:
+            if not repeat:
+                break
+            offset = 0
+
+
+def ieee_eui48_assignment(oui_bits, prefix=0, repeat=False):
+    """
+    IEEE EUI-48 assignment.
+
+    :param oui_bits: The number of OUI bits.
+    :param prefix: The prefix of OUI, for example 0x9a.
+    :param repeat: Whether use repeated identifiers when exhausted.
+    """
+    # Using UUID1 combine with `__file__` to avoid getting the same hash
+    data = uuid.uuid1().hex + __file__
+    data = hashlib.sha256(data.encode()).digest()[:6]
+    sample = 0
+    for num in bytearray(data):
+        sample <<= 8
+        sample |= num
+    bits = 48 - oui_bits
```

Comment:
can we have a macro here and elsewhere ?

Reply:
Yes, better to have one, do you have any suggestion?

Reply:
something like `MAC_ADDRESS_BITS`

Reply:
Instead of using macros (actually python doesn't have built-in macro syntax support), how about making this be:
```python
def ieee_eui_assignment(eui_bits):
    """
    IEEE EUI assignment.

    :param eui_bits: The number of EUI bits.
    """
    def assignment(oui_bits, prefix=0, repeat=False):
        """
        The template of assignment.

        :param oui_bits: The number of OUI bits.
        :param prefix: The prefix of OUI, for example 0x9a.
        :param repeat: Whether use repeated identifiers when exhausted.
        """
        # Using UUID1 combine with `__file__` to avoid getting the same hash
        data = uuid.uuid1().hex + __file__
        data = hashlib.sha256(data.encode()).digest()[:(eui_bits // 8)]
        sample = 0
        for num in bytearray(data):
            sample <<= 8
            sample |= num
        bits = eui_bits - oui_bits
        mask = (1 << bits) - 1
        start = sample & mask
        base = sample ^ start
        if prefix > 0:
            pbits = eui_bits + (-(prefix.bit_length()) // 4) * 4
            pmask = (1 << pbits) - 1
            prefix <<= pbits
            base = prefix | (base & pmask)
    return ieee_eui_generator(base, mask, start, repeat=repeat)
return assignment


ieee_eui48_assignment = ieee_eui_assignment(48)
ieee_eui64_assignment = ieee_eui_assignment(64)
```

Reply:
@luckyh Looks good to me.

Reply:
@balamuruhans updated, thanks.